libs/community/langchain_community/retrievers/sentence_window_retriever.py

```diff
@@ -0,0 +1,184 @@
+from typing import Dict, List
+
+import numpy as np
+from langchain_core.callbacks import CallbackManagerForRetrieverRun
+from langchain_core.documents import Document
+from langchain_core.pydantic_v1 import root_validator
+from langchain_core.retrievers import BaseRetriever
+from langchain_core.vectorstores import VectorStore
+
+# from langchain_pinecone import PineconeVectorStore
+from langchain_community.vectorstores import FAISS, Chroma
+from langchain_community.vectorstores.faiss import (
+    dependable_faiss_import,
+)
+
+
+class SentenceWindowRetriever(BaseRetriever):
+    """`Sentence Window` retriever.
+
+    Sentence Window Retriever dissociates retrieval from generation
+    by appending adjacent chunks to the retrieved chunk to provide
+    additional context to the generation model
+
+    Currently, it supports 2 backends:
+    FAISS, Chroma
+    """
+
+    store: VectorStore
+    """VectorStore containing the chunks of text. Only supports the 
+        above listed databases"""
+
+    k: int = 4
+    """Number of top results to return"""
+
+    window_size: int = 2
+    """Number of adjacent chunks on each side to be added """
+
+    @root_validator
+    def check_database_type(cls, values: Dict) -> Dict:
+        k = values.get("k", -1)
+        window_size = values.get("window_size", -1)
+
+        if k < 0:
+            raise ValueError("The value of k must be greater than 0")
+
+        if window_size < 0:
+            raise ValueError("The value of window_size must be greater than 0")
+
+        return values
+
+    def _get_relevant_documents(
+        self, query: str, *, run_manager: CallbackManagerForRetrieverRun
+    ) -> List[Document]:
+        """Sync implementations for retriever."""
+
+        if type(self.store) == Chroma:
+            return self._sentence_window_retriever_chroma(query=query)
+        # elif type(self.store) == PineconeVectorStore:
+        #     return self._sentence_window_retriever_pinecone(query=query)
+        elif type(self.store) == FAISS:
+            return self._sentence_window_retriever_faiss(query=query)
+        else:
+            raise ValueError(
+                """Only the following databases are currently supported for 
+                the implementation of Sentence Window Retriever : 
+                FAISS, Chroma"""
+            )
+
+    def _sentence_window_retriever_chroma(self, query: str) -> List[Document]:
+        assert isinstance(self.store, Chroma)
+
+        vector = self.store.embeddings.embed_query(query)  # type: ignore
+
+        top_results = self.store._collection.query(
+            query_embeddings=vector, n_results=self.k
+        )
+
+        doc_list = []
+
+        for id, metadata in zip(top_results["ids"][0], top_results["metadatas"][0]):  # type: ignore
```

Comment:
Let's update the code to handle the case of an empty list of results.

Reply:
Got it. Will update for that.

Reply:
The new implementation should be able to handle empty results from the vectorstore.